### ISSUE TITLE
failover destination support in Promtail

### ISSUE THREAD

**Issue by sandeepsukhani** [grafana/loki#8476]
**Is your feature request related to a problem? Please describe.**
When the destination configured in Promtail to push the logs goes down, and if it doesn't come back up quickly enough, then we start losing the logs.

**Describe the solution you'd like**
Add support to Promtail for configuring failover destination.
If the primary destination goes down, Promtail should send logs to the failover destination.

**Describe alternatives you've considered**
An alternative would be using the WAL feature in Promtail when it is ready. It is still a WIP and is being tracked here https://github.com/grafana/loki/issues/8197

**Additional context**
After discussing this internally with other Loki Maintainers, we found some caveats that would require some careful consideration if we go ahead with building this feature. Some of them are:
* Define what a failure is? Should it be just 5xx errors?
* What if the failover destination fails as well?
* Should Promtail keep trying to move back to the Primary destination for pushing the logs? If yes, then how frequently should it do the check?
* Should we support just one failover destination? Does it make sense to have more than one failover destination?

**Comment by adityacs** [comment]
Nothing against this feature addition but just want to share my thoughts. 

1. Adding a failover destination also means more maintenance as we will have one more instance to manage.
2. To not get affected by network glitches(client to destination path), failover instance should be placed in a different network path than the original one. If failover instance is in other network path, there will be some overhead of network configuration for pushing the logs from failover instance to original Loki cluster.

IMHO, WAL is the single solution to most of these problems. It will have some overhead on the client side but should be okay.

**Comment by Eroyi** [comment]
Yeah currently I'm working for a financial giant and trying both Promtail and Fluentd for log tailing. We'll do all kinds of extreme cases chaos test here, one of them is:

1. Set clients.backoff_config.max_retries to 1200 ( 7 days )
2. Shutdown Loki and keep it down
3. Wait 1hrs for Promtail to hoard enough failed batch waiting for resent in memory
4. Restart the Promtail
5. Wait 10in for Promtail to hoard some failed batch waiting for resent
6. Start Loki
7. Check the log integrity

Because Promtail doesn't have a WAL solution, so all the failed batch will be lost while restarting. This could be a common cases especially your server or kubernetes cluster cross multiple region or location.

**Comment by Segflow** [comment]
Promtail was deprecated. Please use [Grafana Alloy](https://grafana.com/docs/alloy/latest/) instead. Closing as obsolete.
